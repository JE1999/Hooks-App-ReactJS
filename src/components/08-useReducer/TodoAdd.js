import React, { memo } from 'react'
import { useForm } from '../../hooks/useForm'


export const TodoAdd = memo(({handleAddTodo}) => {

    const [ {desc}, handleChange, reset ] = useForm({desc: ''});

    const handleSubmit =  e =>{
        e.preventDefault()
        if(desc.trim().length <1){
            return
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: desc,
            done: false
        }

        handleAddTodo(newTodo)
        reset()
    }

    return (
        <>
            <h3>Agregar</h3>
                <form onSubmit={handleSubmit}>
                    <input 
                        type="text"
                        name="desc"
                        className="form-control"
                        placeholder="Agregar..."
                        autoComplete="off"
                        value={desc}
                        onChange={handleChange}
                    />
                    <button
                        type="submit"
                        className="btn btn-block btn-primary"
                    >
                        AGREGAR
                    </button>
                </form>
        </>
    )
})
