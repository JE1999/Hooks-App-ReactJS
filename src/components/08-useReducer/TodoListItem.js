import React, { memo } from 'react'

export const TodoListItem = memo(({todo, index, handleToggle, handleDelete}) => {
    return (
        <li 
            key={todo.id}
            className="list-group-item"
        >
            <p onClick={() => handleToggle(todo.id)} className={`${todo.done && 'text-success'}`} style={{cursor: 'pointer'}}>{index + 1}. {todo.desc}</p>
            <button 
                className="btn btn-sm btn-danger"
                onClick={() => handleDelete(todo.id)}
            >
                Borrar
            </button>
        </li>
    )
})

//Esto se incluye para saber cuantas veces este componente es llamado
//Test
//Solo se incluira si el componente esta rodeado por el memo
TodoListItem.displayName="TodoListItem"
