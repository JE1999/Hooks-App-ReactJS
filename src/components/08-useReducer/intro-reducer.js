const initialState = [{
    id: 1,
    todo: 'Comprar pan',
    done: false
}];

const todoReducer = (state = initialState, action) => {

    if(action?.type === "AGREGAR"){
        return [...state, action.payload]
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
}

const action = {
    type: 'AGREGAR',
    payload: newTodo
}

todos = todoReducer(todos, action)

console.log(todos);