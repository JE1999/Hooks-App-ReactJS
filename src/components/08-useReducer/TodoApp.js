import React, { useReducer, useEffect, useCallback } from 'react'
import { todoReducer } from './todoReducer'
import { TodoList } from './TodoList'
import { TodoAdd } from './TodoAdd'

// const initialState = [{
//     id: new Date().getTime(),
//     desc: 'App React',
//     done: false
// }];

const init = () =>{
    
    return JSON.parse(localStorage.getItem('todos')) || []
    
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'App React',
    //     done: false
    // }]
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    console.log(todos);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos])

    const handleDelete = useCallback(todoId =>{
        const action = {
            type: "DELETE",
            payload: todoId
        }

        dispatch(action)
    },[])

    const handleToggle = useCallback(todoId =>{
        const action = {
            type: "TOGGLE",
            payload: todoId
        }

        dispatch(action)
    },[])

    const handleAddTodo = useCallback(newTodo => {
        dispatch({
            type: 'ADD',
            payload: newTodo
        })
    },[])

    return (
        <div className="container">
            <h1>TodoApp ({todos.length})</h1>
            <hr />

            <TodoAdd 
                handleAddTodo={handleAddTodo}
            />

            <hr />

            <TodoList
                todos={todos}
                handleToggle={handleToggle}
                handleDelete={handleDelete}
            />

        </div>
    )
}
