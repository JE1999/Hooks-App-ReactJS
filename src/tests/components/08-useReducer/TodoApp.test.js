import React from 'react'
import { shallow } from 'enzyme'
import { TodoApp } from '../../../components/08-useReducer/TodoApp'
// import { act } from '@testing-library/react'
// import { demoTodos } from '../../fixtures/demoTodos'



describe('Pruebas en <TodoApp />', () => {
    
    const wrapper = shallow(<TodoApp />)

    test('Debe de mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot()

    })

    // test('Debe de agregar un TODO', () => {
        
    //     //mount aun no es compatible con react 17
    //     const wrapper = mount(<TodoApp />)

    //     act( () =>{
    //         wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[0] )
    //     }) 

    // })

})
