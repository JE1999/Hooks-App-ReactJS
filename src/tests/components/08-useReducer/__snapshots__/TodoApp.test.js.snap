// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> Debe de mostrarse correctamente 1`] = `
<div
  className="container"
>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <TodoAdd
    handleAddTodo={[Function]}
  />
  <hr />
  <Memo()
    handleDelete={[Function]}
    handleToggle={[Function]}
    todos={Array []}
  />
</div>
`;
